File: Satellite_LEO_Magnetorquer_Despin_Controller/eulerseqns2.m
function dydt = eulerseqns2(t,y)
% AE 6356 Spacecraft Attitude
% Assignment 3 Problem 3
% Glenn Lightsey Fall 2024
%
global L % global variable 3x1 control torque
%
% define mass properties
mass=10;     % mass in kg
height=0.34;   % height in m  (x)
width=0.2;   % width in m (y)
depth=0.1;  % depth in m (z)
%
% rectangular prism spacecraft model
Jx=mass/12*(width^2+depth^2);
Jy=mass/12*(height^2+depth^2);
Jz=mass/12*(height^2+width^2);
%
% principal axis inertia tensor
J=[Jx,0,0;0,Jy,0;0,0,Jz];
%
% calculate inertia coefficients
c1=(Jy-Jz)/Jx; 
c2=(Jz-Jx)/Jy; 
c3=(Jx-Jy)/Jz; 
%
w1=y(1); w2=y(2); w3=y(3);
w=[w1;w2;w3];
q1=y(4); q2= y(5); q3=y(6); q4=y(7);
xi=[q4,-q3,q2;q3,q4,-q1;-q2,q1,q4;-q1,-q2,-q3];
qd=0.5*xi*w;
%
dydt = [    c1*w2*w3 + L(1)
            c2*w3*w1 + L(2)
            c3*w1*w2 + L(3)
            qd(1)
            qd(2)
            qd(3)
            qd(4)];

File: Satellite_Architecture/Satellite_Control_Simulation/cpp/src/eulerseqns2.cpp
// eulerseqns2.cpp
#include "eulerseqns2.h"
#include <cmath>

// Definition of global L; must be set in the simulation before each derivative call
std::array<double, 3> L = {0.0, 0.0, 0.0};

std::array<double, 7> eulerseqns2(double t, const std::array<double, 7> &y) {
    // Mass properties (rectangular prism)
    double mass = 10.0;      // kg
    double height = 0.34;    // m (x dimension)
    double width  = 0.20;    // m (y dimension)
    double depth  = 0.10;    // m (z dimension)

    // Principal moments of inertia
    double Jx = mass / 12.0 * (width*width + depth*depth);
    double Jy = mass / 12.0 * (height*height + depth*depth);
    double Jz = mass / 12.0 * (height*height + width*width);

    // Inertia coefficients for Euler equations
    double c1 = (Jy - Jz) / Jx;
    double c2 = (Jz - Jx) / Jy;
    double c3 = (Jx - Jy) / Jz;

    // Unpack state
    double w1 = y[0];
    double w2 = y[1];
    double w3 = y[2];
    std::array<double,3> w = { w1, w2, w3 };
    double q1 = y[3];
    double q2 = y[4];
    double q3 = y[5];
    double q4 = y[6];

    // Construct xi matrix (4x3) to compute quaternion derivative
    // xi * w = q_dot * 2
    // xi = [ [ q4, -q3,  q2 ];
    //        [ q3,  q4, -q1 ];
    //        [-q2,  q1,  q4 ];
    //        [-q1, -q2, -q3 ] ]
    double xi[4][3] = {
        {  q4, -q3,  q2 },
        {  q3,  q4, -q1 },
        { -q2,  q1,  q4 },
        { -q1, -q2, -q3 }
    };

    // Compute q_dot = 0.5 * xi * w
    std::array<double,4> qd;
    for (int i = 0; i < 4; ++i) {
        qd[i] = 0.5 * ( xi[i][0] * w1 + xi[i][1] * w2 + xi[i][2] * w3 );
    }

    // Compute angular acceleration from Euler's equations: w_dot = [c1 w2 w3; c2 w3 w1; c3 w1 w2] + [Lx/Jx; Ly/Jy; Lz/Jz]
    std::array<double,3> wdot;
    wdot[0] = c1 * w2 * w3 + L[0] / Jx;
    wdot[1] = c2 * w3 * w1 + L[1] / Jy;
    wdot[2] = c3 * w1 * w2 + L[2] / Jz;

    // Assemble output derivative vector
    std::array<double,7> dydt;
    dydt[0] = wdot[0];
    dydt[1] = wdot[1];
    dydt[2] = wdot[2];
    dydt[3] = qd[0];
    dydt[4] = qd[1];
    dydt[5] = qd[2];
    dydt[6] = qd[3];

    return dydt;
}